ecosystem/typescript/sdk_v2/src/internal/transaction.ts

```diff
@@ -32,3 +34,167 @@ export async function getGasPriceEstimation(args: { aptosConfig: AptosConfig })
+    aptosConfig,
+    originMethod: "getTransactionByVersion",
+    path: `transactions/by_version/${txnVersion}`,
+  });
+  return data;
+}
+
+export async function getTransactionByHash(args: {
+  aptosConfig: AptosConfig;
+  txnHash: string;
+}): Promise<TransactionResponse> {
+  const { aptosConfig, txnHash } = args;
+  const { data } = await getAptosFullNode<{}, TransactionResponse>({
+    aptosConfig,
+    path: `transactions/by_hash/${txnHash}`,
+    originMethod: "getTransactionByHash",
+  });
+  return data;
+}
+
+export async function isTransactionPending(args: { aptosConfig: AptosConfig; txnHash: string }): Promise<boolean> {
+  const { aptosConfig, txnHash } = args;
+  try {
+    const response = await getTransactionByHash({ aptosConfig, txnHash });
+    return isPendingTransaction(response);
+  } catch (e: any) {
+    if (e?.status === 404) {
+      return true;
+    }
+    throw e;
+  }
+}
+
+/**
+ * Wait for a transaction to move past pending state.
+ *
+ * There are 4 possible outcomes:
+ * 1. Transaction is processed and successfully committed to the blockchain.
+ * 2. Transaction is rejected for some reason, and is therefore not committed
+ *    to the blockchain.
+ * 3. Transaction is committed but execution failed, meaning no changes were
+ *    written to the blockchain state.
+ * 4. Transaction is not processed within the specified timeout.
+ *
+ * In case 1, this function resolves with the transaction response returned
+ * by the API.
+ *
+ * In case 2, the function will throw an ApiError, likely with an HTTP status
+ * code indicating some problem with the request (e.g. 400).
+ *
+ * In case 3, if `checkSuccess` is false (the default), this function returns
+ * the transaction response just like in case 1, in which the `success` field
+ * will be false. If `checkSuccess` is true, it will instead throw a
+ * FailedTransactionError.
+ *
+ * In case 4, this function throws a WaitForTransactionError.
+ *
+ * @param txnHash The hash of a transaction previously submitted to the blockchain.
+ * @param extraArgs.timeoutSecs Timeout in seconds. Defaults to 20 seconds.
+ * @param extraArgs.checkSuccess See above. Defaults to false.
+ * @returns See above.
+ *
+ * @example
+ * ```
+ * const rawTransaction = await this.generateRawTransaction(sender.address(), payload, extraArgs);
+ * const bcsTxn = AptosClient.generateBCSTransaction(sender, rawTransaction);
+ * const pendingTransaction = await this.submitSignedBCSTransaction(bcsTxn);
+ * const transasction = await this.aptosClient.waitForTransactionWithResult(pendingTransaction.hash);
+ * ```
+ */
+export async function waitForTransaction(args: {
+  aptosConfig: AptosConfig;
+  txnHash: string;
+  extraArgs?: { timeoutSecs?: number; checkSuccess?: boolean };
+}): Promise<TransactionResponse> {
+  const { aptosConfig, txnHash, extraArgs } = args;
+  const timeoutSecs = extraArgs?.timeoutSecs ?? DEFAULT_TXN_TIMEOUT_SEC;
+  const checkSuccess = extraArgs?.checkSuccess ?? false;
```

Comment:
imo it makes more sense to default checkSuccess to true, I think most of the time people want to wait for transactions result

Reply:
Definitely agree! Think we should change this to default to true, but keep in mind if you do it needs to be noted/changed in multiple places in the documentation. can generally ctrl + F on `checkSuccess` to see them